### PR TITLE
Portals user obj

### PR DESCRIPTION
**Pull request by squirrelo**
This adds portal checks to the SQL in the user object. It also moves the creation of default analyses to after the user is verified, so we don't clutter up the DB with potentially unused, unverified users.

### CODE REVIEW

File: qiita_db/test/test_user.py [hunk 1]
```diff
@@ -224,38 +219,85 @@ def test_set_info_bad_info(self):
...
+
+            qiita_config.portal = "EMP"
+            self.assertEqual(user.private_analyses, set())
+        finally:
+            qiita_config.portal = self.portal
 
     def test_get_shared_analyses(self):
-        self.assertEqual(self.user.shared_analyses, set([]))
+        user = User('shared@foo.bar')
+        try:
```
Comment (wasade) [inline qiita_db/test/test_user.py]: If the exception is deterministic, can it be caught explicitly? If it is stochastic, then can the test be designed to be deterministic? Same thing with the other try/except blocks

File: qiita_db/user.py [hunk 2]
```diff
@@ -349,55 +356,60 @@ def info(self, info):
 
     @property
     def default_analysis(self):
-        sql = ("SELECT analysis_id FROM qiita.analysis WHERE email = %s AND "
-               "dflt = true")
-        conn_handler = SQLConnectionHandler()
-        return conn_handler.execute_fetchone(sql, [self._id])[0]
-
-    @property
-    def sandbox_studies(self):
```
Comment (josenavas) [inline qiita_db/user.py]: Why did you remove the `sandbox_studies` property?

Comment (squirrelo) [inline qiita_db/user.py]: It was used nowhere in the code, didn't even have a test in the user obj tests, and still references qiita.study_status table.

File: qiita_db/test/test_user.py (hunk 1 above)
Comment (squirrelo) [inline qiita_db/test/test_user.py]: These try/finally are more for making absolutely sure that the portal stays what it should. That way if as you're testing the test fails, it will still reset back to the correct portal no matter what. Safety precaution really.

File: qiita_db/user.py (hunk 2 above)
Comment (josenavas) [inline qiita_db/user.py]: Ok, thanks

File: qiita_db/test/test_user.py (hunk 1 above)
Comment (josenavas) [inline qiita_db/test/test_user.py]: Given that there are multiple functions where this needs to be done, can you add it to the tearDown function? That is the whole idea of that tearDown function. Also, it doesn't hurt if it is executed always. and make the tests more readable.

Comment (squirrelo) [inline qiita_db/test/test_user.py]: Ok.

File: qiita_db/support_files/populate_test_db.sql [hunk 3]
```diff
@@ -440,7 +440,8 @@ INSERT INTO qiita.collection_job (collection_id, job_id) VALUES (1, 1);
...
+INSERT INTO qiita.analysis (email, name, description, dflt, analysis_status_id) VALUES ('test@foo.bar', 'test@foo.bar-dflt-1', 'dflt', true, 1), ('admin@foo.bar', 'admin@foo.bar-dflt-1', 'dflt', true, 1), ('shared@foo.bar', 'shared@foo.bar-dflt-1', 'dflt', true, 1), ('demo@microbio.me', 'demo@microbio.me-dflt-1', 'dflt', true, 1), ('test@foo.bar', 'test@foo.bar-dflt-2', 'dflt', true, 1), ('admin@foo.bar', 'admin@foo.bar-dflt-2', 'dflt', true, 2), ('shared@foo.bar', 'shared@foo.bar-dflt-2', 'dflt', true, 2), ('demo@microbio.me', 'demo@microbio.me-dflt-2', 'dflt', true, 2);
+INSERT INTO qiita.analysis_portal (analysis_id, portal_type_id) VALUES (3, 1), (4, 1), (5, 1), (6, 1), (7, 2), (8, 2), (9, 2), (10, 2);
```
Comment (josenavas) [inline qiita_db/support_files/populate_test_db.sql]: If I understand correctly, what you're doing here is making the default analysis only visible on their portal, right? If so, I think this is correct.

Comment (squirrelo) [inline qiita_db/support_files/populate_test_db.sql]: yup

File: qiita_db/user.py [hunk 4]
```diff
@@ -266,9 +263,9 @@ def verify_code(cls, email, code, code_type):
         elif code_type == 'reset':
             column = 'pass_reset_code'
         else:
-            raise IncompetentQiitaDeveloperError("code_type must be 'create'"
-                                                 " or 'reset' Uknown type "
-                                                 "%s" % code_type)
+            raise IncompetentQiitaDeveloperError(
+                "code_type must be 'create' or 'reset' Uknown type %s" %
```
Comment (josenavas) [inline qiita_db/user.py]: This is not for this PR, and we probably want to open an issue, but, if there are only 2 options `create` and  `reset`, can we represent this as a bool? If `True` means reset, if false means create...

File: qiita_db/user.py [hunk 5]
```diff
@@ -280,15 +277,34 @@ def verify_code(cls, email, code, code_type):
...
+                     WHERE email = %s""".format(cls._table)
             conn_handler.execute(sql, (level, email))
+            # create user default sample holders once verified
+            # create one per portal
+            analysis_sql = """INSERT INTO qiita.analysis
+                (email, name, description, dflt, analysis_status_id)
+                VALUES (%s, %s, %s, %s, 1)"""
+            analysis_args = []
+            sql = "SELECT portal_type_id from qiita.portal_type"
+            for portal in conn_handler.execute_fetchall(sql):
```
Comment (josenavas) [inline qiita_db/user.py]: Can this be done in a list comprehension? They're faster than doing appends to the list.

File: qiita_db/user.py [hunk 6]
```diff
@@ -280,15 +277,34 @@ def verify_code(cls, email, code, code_type):
...
+                analysis_args.append(
+                    (email, '%s-dflt-%d' % (email, portal[0]), 'dflt', True))
+            conn_handler.executemany(analysis_sql, analysis_args)
+
+        if db_code == code:
```
Comment (josenavas) [inline qiita_db/user.py]: This comparison is done up to 3 times in the code (previous if, this one and return statement). Can be computed once and stored in a variable?

**Comment by josenavas** [comment]
Thanks @squirrelo 
I'm done with the review. You've some test failures but the code looks pretty good!

I really like how this is shaping up, with the modifications in the SQL queries. It is a bit tedious, but the end product is going to be awesome!

File: qiita_db/user.py (hunk 4 above)
Comment (squirrelo) [inline qiita_db/user.py]: There could potentially be other codes, for example API keys, so it was left open ended.

File: qiita_db/user.py (hunk 5 above)
Comment (squirrelo) [inline qiita_db/user.py]: done

File: qiita_db/user.py (hunk 6 above)
Comment (squirrelo) [inline qiita_db/user.py]: done

**Comment by squirrelo** [comment]
Comments addressed, ready to go.

**Comment by josenavas** [comment]
:+1:

**Comment by squirrelo** [comment]
This is ready to be merged, and should probably be soon so @josenavas can start working on transactions in #1294

**Comment by josenavas** [comment]
Don't worry @squirrelo, I will solve the merge conflicts later on, otherwise is going to be 2 branches really dirty and one dependent on the other.

File: qiita_db/user.py [hunk 7]
```diff
@@ -280,16 +277,36 @@ def verify_code(cls, email, code, code_type):
             return False
 
         db_code = db_code[0]
+        if not db_code:
+            raise QiitaDBError("No %s code for user %s" %
```
Comment (ElDeveloper) [inline qiita_db/user.py]: Can you add this exception to the `Raises` section of the docstring? Also ... just noticed that `code_type` does not have a description, would you mind adding that as well?

**Comment by ElDeveloper** [comment]
:+1: once the documentation comment is addressed.
